cardano-crypto-tests/src/Test/Crypto/EllipticCurve.hs

```diff
@@ -0,0 +1,224 @@
+    , testProperty "in group" (BLS.inGroup @curve)
+    , testProperty "neg in group" (BLS.inGroup @curve . BLS.neg)
+
+    , testProperty "self-equality" (\(a :: BLS.P curve) -> a === a)
+    , testProperty "double negation" (\(a :: BLS.P curve) -> a === BLS.neg (BLS.neg a))
+    , testProperty "addition associative" (testAssoc (BLS.add :: BLS.P curve -> BLS.P curve -> BLS.P curve))
+    , testProperty "addition commutative" (testCommut (BLS.add :: BLS.P curve -> BLS.P curve -> BLS.P curve))
+    , testProperty "adding negation yields infinity" (testAddNegYieldsInf @curve)
+    , testProperty "round-trip serialization" $
+        testRoundTripEither @(BLS.P curve) BLS.serialize BLS.deserialize
+    , testProperty "round-trip compression" $
+        testRoundTripEither @(BLS.P curve) BLS.compress BLS.uncompress
+    , testProperty "mult by p is inf" $ \(a :: BLS.P curve) ->
+        BLS.isInf (BLS.mult a BLS.scalarPeriod)
+    , testProperty "mult by p+1 is identity" $ \(a :: BLS.P curve) ->
+        BLS.mult a (BLS.scalarPeriod + 1) === a
+    ]
+
+testPT :: String -> TestTree
+testPT name =
+  testGroup name
+    [ testProperty "mult associative"
+        (testAssoc BLS.ptMult)
+    , testProperty "mult commutative"
+        (testCommut BLS.ptMult)
+    , testProperty "self-equality" (\(a :: BLS.PT) -> a === a)
+    , testProperty "self-final-verify" (\(a :: BLS.PT) -> BLS.ptFinalVerify a a)
+    ]
+
+testPairing :: String -> TestTree
+testPairing name =
+  testGroup name
+    [ testProperty "identity" $ \a b ->
+        pairingCheck
+          (a, b)
+          (a, b)
+    , testProperty "simple" $ \a p q ->
+        pairingCheck
+          (BLS.mult p a, q)
+          (p, BLS.mult q a)
+    , testProperty "crossover" $ \a b p q ->
+        pairingCheck
+          (BLS.mult p a, BLS.mult q b)
+          (BLS.mult p b, BLS.mult q a)
+    , testProperty "shift" $ \a b p q ->
+        pairingCheck
+          (BLS.mult p (a * b), q)
+          (BLS.mult p a, BLS.mult q b)
+    , testProperty "three pairings"
+        (\a b p q ->
+            either (const False) id $ do
+                t1 <- BLS.pairing (BLS.mult p a) q
+                t2 <- BLS.pairing p (BLS.mult q b)
+                t3 <- BLS.pairing (BLS.mult p (a + b)) q
+                let tt = BLS.ptMult t1 t2
+                return $ BLS.ptFinalVerify tt t3
+        )
+    ]
+    where
+      pairingCheck (a, b) (c, d) =
+        either (error . show) id $ do
+          p <- BLS.pairing a b
+          q <- BLS.pairing c d
+          return $ BLS.ptFinalVerify p q
+
+testAssoc :: (Show a, Eq a) => (a -> a -> a) -> a -> a -> a -> Property
+testAssoc f a b c =
+  f a (f b c) === f (f a b) c
+
+testCommut :: (Show a, Eq a) => (a -> a -> a) -> a -> a -> Property
+testCommut f a b =
+  f a b === f b a
+
+testNeutral :: (Show a, Eq a) => a -> (a -> a -> a) -> a -> Property
+testNeutral z f x =
+  f x z === x
```

Comment:
This property does not seem to be included in any test tree

Reply:
Good catch. The purpose of this function is to test that a given value `z` is a neutral element (the `mzero` of a `Monoid`) over a given binary operation `f`. This may be a remnant of an earlier iteration, where we did test this property for some operations. It could possibly also be used to replace the existing "mult by p+1 is identity" test as `testNeutral (p+1) BLS.mult`.

Reply:
Removed.